### ISSUE TITLE
Output 3 shortest path routes per OD pair (from `compute_routes()`)

### ISSUE THREAD

**Issue by pavyedav** [cb-cities/ual-sp#4]
This issue has been created to determine the best next steps for implementing `compute_routes()` to output 3 different routes for every OD pair. 

Right now, `compute_routes()` outputs a `std::vector<abm::graph::vertex_t>`. If it needs to output 3 routes per OD now, @kks32 , can you make it do a structure of `std::vector<std::vector<abm::graph::vertex_t>>`? I'm wondering whether this is the best structure to output if we are going to constantly update the edges and calculate the total times of each of these routes again.

@kks32 is suggesting calling `compute_routes(n)` function with a variable `n`, where `n` is the first, second, or n-th route. I believe this is a good idea as well, so as not to hardcode it.

Additional questions include:

1) How much longer will doing this add to runtime?
2) @pavyedav I have made changes to the dynamic_shortest_path branch to have new data structures that take as input the outputs of `compute_routes()`, but this will obviously change as we continue. I'm finding that these data structures and the movement of data may get pretty tricky during the sim, so how do I make this not only work but also work efficiently?

**Comment by pavyedav** [comment]
As of now, `compute_routes()` outputs a vector as follows (as an example): 

`[2 4 6 9 -1 15 18 20 7 -1 12 8 2 9 -1...]`, where each value before a -1 is an edge for that particular person's path. I am envisioning that we just have two more vectors like this to make a 2D vector that is MxN, where M is the number of routes we want to have per OD (3, in our case), and N is the number of total edges for all paths for that run.

As a result, I wrote `createTimeMatrix()` and `updateRouteShareMatrix()`, which assume that the output of `compute_routes()` is `std::vector<std::vector<abm::graph::vertex_t>>`. 

`createTimeMatrix()` creates a 2D vector that contains the total times for each OD's route for M routes (in our case, 3). 

`updateRouteShareMatrix()` create a 2D vector that contains a vector for every OD's probability of taking one of its M routes.

**Comment by kks32** [comment]
Do we remove a few most critical (high centrality index), in the network when computing the second and third route? However, it might mean that for certain people their routes won't change. 

a. Do we remove the top `n` number of highly critical links and then compute the routes?
or
b. Do we remove the most critical link for each person and then try again?

**Comment by pavyedav** [comment]
From Paul: I’m trying to think of ways to reduce the size of this problem somewhat.  How many of the nodes are local street nodes?  I’m guessing that it includes a lot of local street nodes and that is why we have so many OD pairs with only one trip.  If that is the case, perhaps it could help to have a subgraph of the network that contains only non-local streets, reflecting the larger capacity network, and use that subgraph to compute the 3 alternative routes on the smaller network where there should be a lot more overlap in terms of number of trips.  That would also cut down radically the sp computation time since the network would be way smaller.  We could append fixed routes connecting the local street nodes to the nearest higher level node to complete the routes.  These local street node to nearest higher-level street node fixed routes would likely never be congested (?), so we might be able to store those and just look them up as needed to get to the final OD at the full network level.

What do you think, @kks32 ?

**Comment by waddell** [comment]
Rather than using TAZ centroids as is traditionally done in travel modeling to connect OD pairs, we could use nodes on the subgraph of the network containing nodes that are not local street nodes.  It would generally be more nodes than TAZ centroids, and with less spatial distortion for large sparse TAZs, where the TAZ centroid might not be a good approximation of the location of nodes on the network representing collector streets or higher capacity network.

**Comment by kks32** [comment]
@waddell so use network logic and metrics than TAZs, cool!

**Comment by pavyedav** [comment]
The activity demand data is TAZ based, and right now, we randomly assign a node in the TAZ from the large local network. My understanding is then:

1) Select origin/destination nodes in the subgraph closest to the origin/destination nodes in the original graph with local nodes.
2) Compute shortest paths between the subgraph origin/destination nodes (ideally, this should be _much_ faster) (O_{s} to D_{s}), where _s_ is subgraph.
3) Compute shortest paths between the subgraph origin/destinations and the origin/destination nodes in the original graph with local nodes (O_{l} to O_{s} and D_{s} to D_{l}), where _l_ is local.
4) So, in total @kks32, the output of the new shortest path should be: 

- map of every trip O and D with its associated subgraph O and D `LocaltoSubMap` (e.g. if trip X goes from node 1->3->6->14->12, then with form `std::map<std::tuple<graph::vertex_t, graph::vertex_t>, std::map<std::tuple<graph::vertex_t, graph::vertex_t>>`, this would have a map of {1,12} to {6,14}, assuming 6<->14 was the subgraph OD)

- a vector for every single trip from O_{l} to O_{s} `LtoS` (e.g. 1->3->6 above) and D_{s} to D_{l} `StoL` (e.g. 14->12 above)

- a map `subgraphPathsMap` to a vector of 3 vectors of paths (e.g. [6, 7, 14], [6, 12, 58, 19, 14], [6,72,98,42,9,14] in example above) for every O_{s} to D_{s} (ideally of type `std::map<std::tuple<graph::vertex_t, graph::vertex_t>, std::vector<std::vector<float>>>`)

- a path vector `all_paths` (in the current form shown above in comment [2] and using above example `[1 3 6 12 58 19 14 12 -1 15 18 20 7 -1 12 8 2 9 -1...]`, which is in the `indexPathVec` format) assuming default share values, because that's what we will use at the first hour. This is how I'm understanding it - does it make sense?

Then, once it's in the microsim (this is for me @pavyedav, so I'm thinking and writing out loud here [please disregard if you want to]):

5) When I get new edge speeds at every increment, I need to update the weights (times) of the edges.
6) Look up matrix `S` and determine the new total times of each OD pair's 3 routes, and write them to a matrix `T`. (maybe I'll make `S` and `T` maps of `std::map<std::tuple<graph::vertex_t, graph::vertex_t>, std::vector<float>>`? or I'll stick with the matrix version because it may be faster, but it's also prone to more error because it may get screwed up by indices, etc.) Prelim version exists as a matrix, but requires reworking.
7) Take in map `T` and calculate route shares in map `RS` (same size as map `T`). Prelim version exists as a matrix, but requires reworking.
8) Filter to the upcoming trips for the next hour and that contain all O_{s} to D_{s}, and then apply the route shares to all of these trips by replacing the current path vectors with the new path vectors.
9) Convert to indexPathVec.
10) Transfer indexPathVec back to the GPU via CUDA.
11) Rinse and repeat until sim is done.

I'm unsure just how long (5)-(8) will take. I know (9) takes several minutes right now, so this would have to continue to be the case since we must run that conversion every increment we capture new edge values.

**Comment by pavyedav** [comment]
I have made a ton of additions and changes to the microsim code based on the above description / template. I just pushed that change on the `dynamic_shortest_path` branch.

Please let me know the status of the SP change and if you can produce the specific outputs I highlighted above.

**Comment by kks32** [comment]
@pavyedav 

I have a few questions:

1. How do we identify the central TAZ node? Would this be specified?

2. When updating the weights, are we going to update the weights of all the edges or only the major roads?

3. If we update the weight of all the edges, then shouldn't we compute the whole route again?

**Comment by pavyedav** [comment]
1. Per my understanding, we don't need a TAZ central node. We need just a subgraph of the smaller network without tertiary roads. Then, the nodes in the subgraph closest to the O/D nodes in the larger graph (this 'distance' or 'time' would need to be computed) are the ones used to determine the smaller computation of routes between those subgraph nodes. I can provide this smaller, primary road network.

2. We are updating the weights of all the edges in the current implementation, but I suppose we don't *have* to update the local road weights because they will be the same for all the routes for that traveler. However, see below.

3. By updating the weights of all the edges, then we get a more accurate travel time for each traveler, even though those local road edges are the same for every traveler's 3 routes. The major updates are to the major road/subgraph roads, which will then help determine which route of the 3 the travelers will take, but the weight update for all edges would still be useful to get a 'dynamic' travel time.

Not sure whether this makes sense, but this is my understanding. Let me know!

**Comment by kks32** [comment]
Proposal: Once we have a sub-network of major roads, we identify all the minor nodes and have an associated major node (on the subnetwork) and the distance this node. So there is only 1 major node associated with each minor road. Once this is done, we can then do the routing dynamically on the sub-network. Does that sound reasonable?

**Comment by pavyedav** [comment]
This makes sense and sounds reasonable, but I'm going to try to use an example to understand what you're saying a bit better.

Say we have a path for an OD pair in the large network from node 1->3->6->58->19->14->12. Then, we create the sub-network, and the only nodes from that path that exist in the sub-network now are 6 and 14. All nodes from the large network that are closest to 6 will be mapped to 6, and all nodes closest to 14 will be mapped to 14, etc. for every node in the large network. 

Will this mapping include the path to take from the large graph node to the sub-network node? (e.g. if node 1 was mapped to node 6 from the subgraph, will we have 1->3->6 associated with node 1 so that it's a quick lookup?) And then, we do dynamic routes from 6->14 on the subgraph where it could be 6->58->19->14, 6->42->75->14, or 6->10->81->29->14 as the three possible paths for that OD pair, and then the rest of the route is just taken from the large subgraph mapping you're talking about?

Is this the correct interpretation?

**Comment by kks32** [comment]
Yes, if we consider a whole path as `1->3->6->58->19->14->12`, and the subnetwork path is `6->58->19->14`, we store the closes node to the minor roads and their paths, for quick lookups:
```
1: 6, <1,3,6>
3: 6, <3, 6>
12:14, <12, 14>
```

Then we do dynamic routing between node 6 and node 14. to get `6->58->19->14`. If we do this, then is there a reason to store 3 alternatives? The reason I'm not sure how to do alternatives is that there might not be 3 alternatives and which edge should we remove to get the second and third routes. However, if it's a smaller subnetwork we could potentially do dynamic routing. What do you think? This shouldn't change anything in your code, but only the sp code changes. There is an upfront cost of finding the path to the closest main road node.

**Comment by pavyedav** [comment]
Yup, this makes sense.

Re: whether we need to store 3 alternatives now, I definitely see the issue with figuring out 3 different routes. When you say dynamic routing here though, what would that mean in practice? If this traveler goes from node 1 to node 12, and we now have a subnetwork, we could see what node they've reached during simulation and then go back to the mapping in the subgraph and see if a better route exists in the subgraph? I guess an example would be nice here again to see how it would work.

And iirc, the suggestion with 3 routes was basically a way to have some of the travelers who depart later take different routes for the basically the same OD pair, as a way to add some more dynamism to the movements, so not sure if this is the best approach anyway.

**Comment by kks32** [comment]
If we have a subnetwork, and assuming the shortest path compute is faster on a smaller graph, couldn't we compute the route dynamically in computation time?

**Comment by pavyedav** [comment]
As in, instead of having the routes ready for every single trip, we compute the route within the simulator when each traveler departs, taking into account the new edge weights due to other vehicles on the edge?

**Comment by kks32** [comment]
Yes, we can try to see if we can dynamically reroute in a reasonable time.

**Comment by pavyedav** [comment]
Yeah, interesting. I'm definitely open to trying it like this, but before we try that, should we have an idea of how long calculating a route would be in the smaller subgraph? The smaller subgraph has about 30K nodes and 65K edges, iirc. 3.1M routes on the large 225K node, 550K edge graph took 62 minutes, so each route took approximately 20 _micro_seconds, so assuming some level of linearity where our graph is 7x smaller, are we expecting approximately 3 _micro_seconds? So, we're saying that computing all the routes would take about 62/7 = 8-9 mins total. Is this making sense? If so, then that sounds fantastic. 

Let me know if I'm not thinking about the performance correctly.

**Comment by kks32** [comment]
Could you send me a graph, and I can run some benchmarks?

**Comment by pavyedav** [comment]
Yup. I'll send you both the full graph (nodes and edges), and the subgraph (nodes and edges), with the demand.
